validator_client/src/validator_duty.rs

```diff
@@ -53,58 +55,79 @@ impl ValidatorDuty {
         beacon_node: &BeaconNodeHttpClient,
         current_epoch: Epoch,
         request_epoch: Epoch,
-        pubkey: PublicKey,
-    ) -> Result<ValidatorDuty, String> {
-        let pubkey_bytes = PublicKeyBytes::from(&pubkey);
+        mut known_pubkeys: Vec<(PublicKey, u64)>,
```

Comment:
I have an alternate pattern, feel free to ignore if you like.

Instead of `known_pubkeys` and `unknown_pubkeys`, just have a `pubkeys: &[(PublicKey, Option<u64>)]`. Then, before the `for` loop on L64, define an `indices = Vec::with_capacity(pubkeys.len())` and push into it whenever you discover a validator index.

This removes `known_pubkeys` from this function and the caller, replacing it with `indicies` which is in this function only. It also removes an iteration on L85.

Reply:
Ok, I added this in, but needed an additional way to track the pubkeys of the validator indices we query for. They become necessary when creating empty duties for validators we query for but get no results for. Let me know what you think.